synapse/storage/databases/main/cache.py

```diff
@@ -240,19 +241,26 @@ def _invalidate_caches_for_event(
             self._invalidate_local_get_event_cache(redacts)
             # Caches which might leak edits must be invalidated for the event being
             # redacted.
-            self.get_relations_for_event.invalidate((redacts,))
-            self.get_applicable_edit.invalidate((redacts,))
+            self._attempt_to_invalidate_cache("get_relations_for_event", (redacts,))
+            self._attempt_to_invalidate_cache("get_applicable_edit", (redacts,))
 
         if etype == EventTypes.Member:
             self._membership_stream_cache.entity_has_changed(state_key, stream_ordering)
-            self.get_invited_rooms_for_local_user.invalidate((state_key,))
+            self._attempt_to_invalidate_cache(
+                "get_invited_rooms_for_local_user", (state_key,)
+            )
 
         if relates_to:
-            self.get_relations_for_event.invalidate((relates_to,))
-            self.get_aggregation_groups_for_event.invalidate((relates_to,))
-            self.get_applicable_edit.invalidate((relates_to,))
-            self.get_thread_summary.invalidate((relates_to,))
-            self.get_thread_participated.invalidate((relates_to,))
+            self._attempt_to_invalidate_cache("get_relations_for_event", (relates_to,))
+            self._attempt_to_invalidate_cache(
+                "get_aggregation_groups_for_event", (relates_to,)
+            )
+            self._attempt_to_invalidate_cache("get_applicable_edit", (relates_to,))
+            self._attempt_to_invalidate_cache("get_thread_summary", (relates_to,))
+            self._attempt_to_invalidate_cache("get_thread_participated", (relates_to,))
+            self._attempt_to_invalidate_cache(
+                "get_mutual_event_relations_for_rel_type", (relates_to,)
+            )
```

Comment:
This is significantly greedier than the old version which only invalidated caches based on the relation type. Not sure if this matters much, but we know ahead of time *which* of these *might* need to be invalidated.

Reply:
(Also it is quite annoying that you can no longer jump to these definitions via PyCharm or anything since they're strings now. 😢 )

Reply:
I wonder if this function should take `relation` instead of `relatest_to` and it could evaluate the lost conditionals to reduce the invalidation amount?

>  (Also it is quite annoying that you can no longer jump to these definitions via PyCharm or anything since they're strings now. 😢 )

:( not ideal at all, nor do I have any ideas on how to solve this unfortunately! If all the DB classes were fully shared this could just assume each method existed and call them direct...

Reply:
> I wonder if this function should take `relation` instead of `relatest_to` and it could evaluate the lost conditionals to reduce the invalidation amount?

Maybe, I'm not sure it is a huge deal or not, but definitely fits in the bucket of "code that doesn't need to run". `_attempt_to_invalidate_cache` sounds scary (and looking at it, it seems to use quite a few `getattr` calls, which aren't known for optimizations...)

> >  (Also it is quite annoying that you can no longer jump to these definitions via PyCharm or anything since they're strings now. 😢 )
> 
> :( not ideal at all, nor do I have any ideas on how to solve this unfortunately! If all the DB classes were fully shared this could just assume each method existed and call them direct...

I think ideally it would be nice if we referred to everything by the store it is on, instead of via a god object. Like the changes we did with the config objects. I think this is doable, but a lot of work. 😢 See #11165

Anyway -- I don't think there's anything actionable to do here, just wanted to note my concerns down for posterity.